subsys/bluetooth/mesh/proxy_cli.c

```diff
@@ -183,43 +179,112 @@ static bool has_net_id(struct bt_mesh_subnet *sub, void *user_data)
 		}
 	}
 
-	if (!memcmp(sub->keys[0].net_id, res->net_id, 8) ||
-	    (bt_mesh_subnet_has_new_key(sub) &&
-	     !memcmp(sub->keys[1].net_id, res->net_id, 8))) {
-		res->srv = srv;
+	/* If net_id is NULL we already know that the networks match */
+	if (!net_id || !memcmp(sub->keys[0].net_id, net_id, 8) ||
+	    (bt_mesh_subnet_has_new_key(sub) && !memcmp(sub->keys[1].net_id, net_id, 8))) {
+		*p_srv = srv;
 		return true;
 	}
 
 	return false;
 }
 
-void bt_mesh_proxy_cli_adv_recv(const struct bt_le_scan_recv_info *info,
-				struct net_buf_simple *buf)
-{
+struct find_net_id {
 	uint8_t type;
+
+	union {
+		const uint8_t *net_id;
+		struct {
+			const uint8_t *hash;
+			const uint8_t *rand;
+		} priv;
+	} data;
+
+	struct bt_mesh_proxy_server *srv;
+};
+
+static bool has_net_id(struct bt_mesh_subnet *sub, void *user_data)
+{
+	int err;
+	struct find_net_id *res = user_data;
+	uint8_t *net_id = NULL;
+
+	if (res->type == BT_MESH_ID_TYPE_NET) {
+		net_id = (uint8_t *)res->data.net_id;
+		goto end;
+	}
+
+	/* Additional handling for BT_MESH_ID_TYPE_PRIV_NET msg type */
+	uint8_t tmp[16] = {0};
+	uint8_t idx = bt_mesh_subnet_has_new_key(sub) ? 1 : 0;
+
+	memcpy(&tmp[0], sub->keys[idx].net_id, 8);
+	memcpy(&tmp[8], res->data.priv.rand, 8);
+	err = bt_mesh_encrypt(&sub->keys[idx].identity, tmp, tmp);
```

Comment:
It is better to keep separate incoming and outgoing buffers for crypto functions, since mesh depends on crypto library implementation.